src/storage/heap_file.c

```diff
@@ -25090,20 +25093,102 @@ heap_get_visible_version_with_repl_desync (THREAD_ENTRY * thread_p, HEAP_GET_CON
 	  break;
 	}
       assert (is_passive_transaction_server ());
+      assert (context->scan_cache != nullptr);
+
       // Handle the page desynchronization error
-      if (context->home_page_watcher.pgptr)
-	{
-	  /* Unfix home page. */
-	  pgbuf_ordered_unfix (thread_p, &context->home_page_watcher);
-	}
+      //
 
-      if (context->fwd_page_watcher.pgptr != NULL)
+      // clean - saving state if required
+      //
+      heap_clean_get_context (thread_p, context);
+
+      // also unfix what is in cache watcher; saving the state for a future re-fix after the wait
+      //
+      VPID saved_vpid VPID_INITIALIZER;
+      PGBUF_LATCH_MODE saved_latch_mode = PGBUF_NO_LATCH;
+      bool restore_scan_cache_page_watcher = false;
+      if (context->scan_cache != nullptr && context->scan_cache->page_watcher.pgptr != nullptr)
 	{
-	  /* Unfix forward page. */
-	  pgbuf_ordered_unfix (thread_p, &context->fwd_page_watcher);
+	  assert (context->scan_cache->cache_last_fix_page);
+
+	  pgbuf_ordered_unfix_and_save_for_refix (thread_p, &context->scan_cache->page_watcher,
+						  saved_vpid, saved_latch_mode);
+	  assert (!VPID_ISNULL (&saved_vpid));
+	  assert (PGBUF_LATCH_READ == saved_latch_mode);
+	  restore_scan_cache_page_watcher = true;
 	}
+
+      assert (PGBUF_IS_CLEAN_WATCHER (&context->home_page_watcher));
+      assert (PGBUF_IS_CLEAN_WATCHER (&context->fwd_page_watcher));
+      assert (context->scan_cache == nullptr || PGBUF_IS_CLEAN_WATCHER (&context->scan_cache->page_watcher));
+
+      // wait for replication to catch-up
+      //
       constexpr VPID null_vpid = VPID_INITIALIZER;
       pgbuf_wait_for_replication (thread_p, &null_vpid);
+
+      // re-fix scan cache page watcher, if unfixed previously
+      //
+      if (restore_scan_cache_page_watcher)
+	{
+	  const int error_code = pgbuf_ordered_fix (thread_p, &saved_vpid, OLD_PAGE_IF_IN_BUFFER_OR_IN_TRANSIT,
```

Comment:
@razvanradulescu 
After some more though, I changed to use the `OLD_PAGE_IF_IN_BUFFER_OR_IN_TRANSIT` for the re-fix.
I also added an extensive comment below explaining the rationale.